src/storage/caching_file_system.cpp

```diff
@@ -90,30 +90,42 @@ BufferHandle CachingFileHandle::Read(data_ptr_t &buffer, const idx_t nr_bytes, c
 
 	// Try to read from the cache, filling overlapping_ranges in the process
 	vector<shared_ptr<CachedFileRange>> overlapping_ranges;
-	result = TryReadFromCache(buffer, nr_bytes, location, overlapping_ranges);
+	idx_t start_location_of_next_range = DConstants::INVALID_INDEX;
+	result = TryReadFromCache(buffer, nr_bytes, location, overlapping_ranges, start_location_of_next_range);
 	if (result.IsValid()) {
 		return result; // Success
 	}
 
+	idx_t new_nr_bytes = nr_bytes;
+	if (start_location_of_next_range != DConstants::INVALID_INDEX) {
+		if (start_location_of_next_range < location + 2 * nr_bytes &&
+		    start_location_of_next_range < location + nr_bytes + 1000000) {
+			// Grow the range from location to start_location_of_next_range, so that to fill gaps in the cached ranges
+			new_nr_bytes = start_location_of_next_range - location;
+		}
+	}
```

Comment:
Let's make `start_location_of_next_range` an `optional_idx`, and `nr_bytes` non-`const`, then we can do something like this:
```C++
if (start_location_of_next_range.IsValid()) {
	static constexpr idx_t MINIMUM_ADDED_BYTES = 10000;
	static constexpr idx_t MAXIMUM_ADDED_BYTES = 1000000;

	const auto added_bytes_bound = ClampValue(MINIMUM_ADDED_BYTES, nr_bytes, MAXIMUM_ADDED_BYTES);
	const auto new_nr_bytes = start_location_of_next_range - location;
	if (new_nr_bytes <= nr_bytes + added_bytes_bound) {
		nr_bytes = new_nr_bytes;
	}
}
```

Reply:
Thanks! I will fix this up and undraft

Reply:
Discussed over Discord but adding here so we don't forget: `start_location_of_next_range` can be initialized with the file size (if available)